### PR TITLE
scylla-macros: increase hygiene

### PR DESCRIPTION
**Pull request by wmitros**
While trying to make my macros at https://github.com/wmitros/scylla-rust-udf/pull/1 more hygienic, I noticed that I also need the macros here to be more hygienic.
I'm posting this mostly to find out if I'm heading in the right direction with the changes, although this patch does pass tests if it turns out being close to a state ready for merging

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
~~- [ ] I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Comment by cvybhu** [comment]
Please rebase on top of latest `main` to fix the CI checks

File: scylla-macros/src/into_user_type.rs [hunk 1]
```diff
@@ -14,20 +14,20 @@ pub fn into_user_type_derive(tokens_input: TokenStream) -> TokenStream {
...
         }
     });
 
     let generated = quote! {
         impl #impl_generics scylla::frame::value::Value for #struct_name #ty_generics #where_clause {
-            fn serialize(&self, buf: &mut Vec<u8>) -> std::result::Result<(), scylla::frame::value::ValueTooBig> {
+            fn serialize(&self, buf: &mut ::std::vec::Vec<::core::primitive::u8>) -> ::std::result::Result<(), scylla::frame::value::ValueTooBig> {
                 use scylla::frame::value::{Value, ValueTooBig};
                 use scylla::macros::BufMut;
                 use ::std::convert::TryInto;
```
Comment (piodul) [inline scylla-macros/src/into_user_type.rs]: Would importing the primitive types, just like `BufMut/TryInto/etc.` here work?

```rust
use ::std::primitive::{i32, usize};
```

File: scylla-macros/src/from_row.rs [hunk 2]
```diff
@@ -22,8 +22,8 @@ pub fn from_row_derive(tokens_input: TokenStream) -> TokenStream {
                     .unwrap(); // vals_iter size is checked before this code is reached, so
                                // it is safe to unwrap
 
-                <#field_type as FromCqlVal<Option<CqlValue>>>::from_cql(col_value)
-                    .map_err(|e| FromRowError::BadCqlVal {
+                <#field_type as scylla::cql_to_rust::FromCqlVal<::std::option::Option<scylla::frame::response::result::CqlValue>>>::from_cql(col_value)
+                    .map_err(|e| scylla::cql_to_rust::FromRowError::BadCqlVal {
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-macros/src/from_row.rs]: Ah, I think there are some tests in scylla-cql that need to use the macros and do `use crate as scylla` in order for the macros to work... I'm not sure what would be the best way to approach this at the moment, so I think we can make an exception for the `scylla` crate at the moment.

**Review by piodul — CHANGES_REQUESTED** [review]
I think the amount of changes that need to be done here can be reduced. Instead of fully qualifying everything, which is tedious, you could import some items with the `use` directive. As long as you are importing in a scope created by the macro which does not pollute the global namespace, it should be fine. For example, importing inside a method body created by the derive macro is OK, but importing items just before the `impl` block is not.

It would also be nice to have a test that verifies that the macros are indeed hygienic. It shouldn't be hard to add, just create an integration test (in scylla-cql or scylla crate) that uses `#[no_implicit_prelude]` and use the derive macros there. The `scylla` crate would be an exception, you will have to import it. Ref: https://gist.github.com/Kestrer/8c05ebd4e0e9347eb05f265dfb7252e1#test-your-macro-with-no_implicit_prelude

File: scylla-macros/src/from_row.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wmitros) [reply scylla-macros/src/from_row.rs]: Yes, this is an issue that I also had

File: scylla-macros/src/into_user_type.rs (hunk 1 above)
Comment (wmitros) [reply scylla-macros/src/into_user_type.rs:26]: Yes, though I used `core` instead of `std` for no particular reason

**Comment by wmitros** [comment]
In the last rebase, I replaced the full paths with `use` statements in every place it shortened the code, adjusted the `get_path` function so that it does not allow renaming the crate multiple times, and added a `no_implicit_prelude` test for renaming and just using the derive macros

File: scylla-macros/src/parser.rs [hunk 3]
```diff
@@ -17,3 +18,48 @@ pub(crate) fn parse_named_fields<'a>(
...
+            }
+            Err(err) => {
+                return Err(err);
+            }
+        }
+    }
+    match this_path {
+        Some(path) => Ok(path),
+        None => Ok(quote::quote!(scylla)),
+    }
```
Comment (piodul) [inline scylla-macros/src/parser.rs:64]: Nit: `this_path.unwrap_or_else(|| quote::quote!(scylla));`

File: scylla/tests/hygiene.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    let mut buf = Vec::<u8>::new();
+    test_struct.serialize(&mut buf).unwrap();
+    let mut buf_assert = Vec::<u8>::new();
+    let tuple_with_same_layout: (i32,) = (16,);
+    tuple_with_same_layout.serialize(&mut buf_assert).unwrap();
+    assert_eq!(buf, buf_assert);
+
+    let sv = test_struct.serialized().unwrap().into_owned();
+    let sv2 = tuple_with_same_layout.serialized().unwrap().into_owned();
+    assert_eq!(sv, sv2,);
```
Comment (piodul) [inline scylla/tests/hygiene.rs:63]: Nit: comma after `sv2` is unnecessary

File: scylla/tests/hygiene.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    use ::std::assert_eq;
+    use ::std::option::Option::Some;
+    use ::std::vec::Vec;
+    use _scylla::_macro_internal::{CqlValue, Row, Value, ValueList};
+    use _scylla::cql_to_rust::FromRow;
+
+    let test_struct = TestStruct { a: 16 };
+    fn get_row() -> Row {
+        Row {
+            columns: ::std::vec![Some(CqlValue::Int(16)),],
```
Comment (piodul) [inline scylla/tests/hygiene.rs:44]: Nit: comma before the `]` is unnecessary

File: scylla-macros/src/parser.rs [hunk 6]
```diff
@@ -17,3 +18,48 @@ pub(crate) fn parse_named_fields<'a>(
...
+                    return Err(syn::Error::new_spanned(
+                        &meta_name_value.lit,
+                        "the `scylla_crate` attribute should be a string literal",
+                    ));
+                }
+            }
+            Ok(other) => {
+                if !other.path().is_ident("scylla_crate") {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (wmitros) [reply scylla-macros/src/parser.rs]: Ok

File: scylla/tests/hygiene.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
+#![no_implicit_prelude]
+
+extern crate scylla as _scylla;
```
[the remark this answers is not in the document]
Comment (wmitros) [reply scylla/tests/hygiene.rs]: I added a test for the `scylla_cql` crate, but I'm not completely sure it was what you expected, please take a look

File: scylla-macros/src/parser.rs (hunk 3 above)
Comment (wmitros) [reply scylla-macros/src/parser.rs]: Ok(Ok)

File: scylla/tests/hygiene.rs (hunk 4 above)
Comment (wmitros) [reply scylla/tests/hygiene.rs]: Ok

File: scylla/tests/hygiene.rs (hunk 5 above)
Comment (wmitros) [reply scylla/tests/hygiene.rs]: Ok, I wonder why `cargo fmt` didn't catch that

**Review by piodul — APPROVED** [review]

**Pull request merged**